File: src/chain-observers/providers/block-header-provider.ts
import { BlockHeaderClassifier } from "../classifiers/block-header-classifier";
import type { IBlockHeaderData } from "../classifiers/block-header-classifier";
import type { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import type { TProviderEvaluationContext } from "../factories/data-evaluation-context";
import { ProviderBase } from "./provider-base";

export interface IBlockHeaderProviderData {
  block: IBlockHeaderData;
};

export class BlockHeaderProvider extends ProviderBase {
  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [
      BlockHeaderClassifier
    ]
  }

  public async provide(data: TProviderEvaluationContext): Promise<IBlockHeaderProviderData> {
    const blockHeader = await data.get(BlockHeaderClassifier);

    return {
      block: {
        ...blockHeader
      }
    };
  }
}

File: python/workerbee/chain_observers/providers/block_header_provider.py
"""BlockHeaderProvider — provides block header data.

Mirrors src/chain-observers/providers/block-header-provider.ts.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, cast

from ..classifiers.block_header_classifier import BlockHeaderClassifier
from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
from .provider_base import ProviderBase

if TYPE_CHECKING:
    from ..factories.data_evaluation_context import DataEvaluationContext
    from ..payloads import BlockHeaderData, BlockHeaderPayload


class BlockHeaderProvider(ProviderBase):
    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        return [BlockHeaderClassifier]

    async def provide(self, data: DataEvaluationContext) -> BlockHeaderPayload:
        block_header = await data.get(BlockHeaderClassifier)

        return {
            "block": cast("BlockHeaderData", {**block_header}),
        }
